README
B-Hooks-OP-Annotation version 0.43
==================================

This module allows XS modules to annotate and delegate hooked OPs.
See the POD for more details.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  ExtUtils::Depends

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009 by chocolateboy

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension B::Hooks::OP::Annotation.

0.43  Fri Feb 27 00:40:07 2009
    - #43682: fix typo in function name (thanks Taro Nishino)

0.42  Thu Feb 26 01:44:26 2009
    - fix deletion

0.41  Wed Feb 25 23:18:35 2009
    - fix hash function

0.40  Wed Feb 25 20:57:42 2009
    - replace khash.h with optable.h
    - addded op_annotate - void version of op_annotation_new
    - gcc fix (thanks BinGOs)

0.32  Sun Feb 22 01:19:50 2009
    - fix metadata
    - fix metadata info in POD

0.31  Sun Feb 15 02:43:49 2009
    - Makefile.PL fixlet

0.30  Fri Feb 13 16:02:55 2009
    - API change
      - don't expose annotation destructor
      - rename op_annotation_set => op_annotation_new and return the new annotation
      - destroy clobbered annotation if the OP has already been annotated
      - s/ppaddr/op_ppaddr/ to appease ppport.h
    - don't issue a warning if data is supplied without a destructor

0.21  Sat Feb 7 03:50:37 2009
    - portability fix
    - fix destructor prototypes
    - POD fixes

0.20  Thu Feb 5 01:19:47 2009
    - API overhaul - use a pull interface
    - replace linked list with khash: http://tinyurl.com/3kkclt

0.10  Wed Feb 4 04:26:02 2009
    - API cleanup
    - simplify/optimize groups

0.01  Wed Feb 4 01:40:17 2009
    - original version; created by h2xs 1.23 with options -A -b5.8.0 --name=B::Hooks::OP::Annotation


khash.h
/* The MIT License

   Copyright (c) 2008, by Attractive Chaos <attractivechaos@aol.co.uk>

   Permission is hereby granted, free of charge, to any person obtaining
   a copy of this software and associated documentation files (the
   "Software"), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to
   permit persons to whom the Software is furnished to do so, subject to
   the following conditions:

   The above copyright notice and this permission notice shall be
   included in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
   SOFTWARE.
*/

/*
  An example:

#include "khash.h"
KHASH_MAP_INIT_INT(32, char)
int main() {
    int ret, is_missing;
    khiter_t k;
    khash_t(32) *h = kh_init(32);
    k = kh_put(32, h, 5, &ret);
    if (!ret) kh_del(32, h, k);
    kh_value(h, k) = 10;
    k = kh_get(32, h, 10);
    is_missing = (k == kh_end(h));
    k = kh_get(32, h, 5);
    kh_del(32, h, k);
    for (k = kh_begin(h); k != kh_end(h); ++k)
    if (kh_exist(h, k)) kh_value(h, k) = 1;
    kh_destroy(32, h);
    return 0;
}
*/

/*
  2008-09-19 (0.2.3):

    * Corrected the example
    * Improved interfaces

  2008-09-11 (0.2.2):

    * Improved speed a little in kh_put()

  2008-09-10 (0.2.1):

    * Added kh_clear()
    * Fixed a compiling error

  2008-09-02 (0.2.0):

    * Changed to token concatenation which increases flexibility.

  2008-08-31 (0.1.2):

    * Fixed a bug in kh_get(), which has not been tested previously.

  2008-08-31 (0.1.1):

    * Added destructor
*/

#ifndef __AC_KHASH_H
#define __AC_KHASH_H

#define AC_VERSION_KHASH_H "0.2.2"

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

typedef uint32_t khint_t;
typedef khint_t khiter_t;

#define __ac_HASH_PRIME_SIZE 32
static const uint32_t __ac_prime_list[__ac_HASH_PRIME_SIZE] =
{
  0ul,          3ul,          11ul,         23ul,         53ul,
  97ul,         193ul,        389ul,        769ul,        1543ul,
  3079ul,       6151ul,       12289ul,      24593ul,      49157ul,
  98317ul,      196613ul,     393241ul,     786433ul,     1572869ul,
  3145739ul,    6291469ul,    12582917ul,   25165843ul,   50331653ul,
  100663319ul,  201326611ul,  402653189ul,  805306457ul,  1610612741ul,
  3221225473ul, 4294967291ul
};

#define __ac_isempty(flag, i) ((flag[i>>4]>>((i&0xfU)<<1))&2)
#define __ac_isdel(flag, i) ((flag[i>>4]>>((i&0xfU)<<1))&1)
#define __ac_iseither(flag, i) ((flag[i>>4]>>((i&0xfU)<<1))&3)
#define __ac_set_isdel_false(flag, i) (flag[i>>4]&=~(1ul<<((i&0xfU)<<1)))
#define __ac_set_isempty_false(flag, i) (flag[i>>4]&=~(2ul<<((i&0xfU)<<1)))
#define __ac_set_isboth_false(flag, i) (flag[i>>4]&=~(3ul<<((i&0xfU)<<1)))
#define __ac_set_isdel_true(flag, i) (flag[i>>4]|=1ul<<((i&0xfU)<<1))

static const double __ac_HASH_UPPER = 0.77;

/* chocolateboy 2009-02-13 - removed inline - leave it to the compiler (which may not support C99 anyway) */

/* BEGIN_MACRO */
#define KHASH_INIT(name, khkey_t, khval_t, kh_is_map, __hash_func, __hash_equal)                                    \
    typedef struct OPAnnotationGroupImpl {                                                                          \
        khint_t n_buckets, size, n_occupied, upper_bound;                                                           \
        uint32_t *flags;                                                                                            \
        khkey_t *keys;                                                                                              \
        khval_t *vals;                                                                                              \
    } kh_##name##_t;                                                                                                \
    static kh_##name##_t *kh_init_##name() {                                                                 \
        return (kh_##name##_t*)calloc(1, sizeof(kh_##name##_t));                                                    \
    }                                                                                                               \
    static void kh_destroy_##name(kh_##name##_t *h)                                                          \
    {                                                                                                               \
        if (h) {                                                                                                    \
            free(h->keys); free(h->flags);                                                                          \
            free(h->vals);                                                                                          \
            free(h);                                                                                                \
        }                                                                                                           \
    }                                                                                                               \
    static void kh_clear_##name(kh_##name##_t *h)                                                            \
    {                                                                                                               \
        if (h && h->flags) {                                                                                        \
            memset(h->flags, 0xaa, ((h->n_buckets>>4) + 1) * sizeof(uint32_t));                                     \
            h->size = h->n_occupied = 0;                                                                            \
        }                                                                                                           \
    }                                                                                                               \
    static khint_t kh_get_##name(kh_##name##_t *h, khkey_t key)                                              \
    {                                                                                                               \
        if (h->n_buckets) {                                                                                         \
            khint_t inc, k, i, last;                                                                                \
            k = __hash_func(key); i = k % h->n_buckets;                                                             \
            inc = 1 + k % (h->n_buckets - 1); last = i;                                                             \
            while (!__ac_isempty(h->flags, i) && (__ac_isdel(h->flags, i) || !__hash_equal(h->keys[i], key))) {     \
                if (i + inc >= h->n_buckets) i = i + inc - h->n_buckets;                                            \
                else i += inc;                                                                                      \
                if (i == last) return h->n_buckets;                                                                 \
            }                                                                                                       \
            return __ac_iseither(h->flags, i)? h->n_buckets : i;                                                    \
        } else return 0;                                                                                            \
    }                                                                                                               \
    static void kh_resize_##name(kh_##name##_t *h, khint_t new_n_buckets)                                    \
    {                                                                                                               \
        uint32_t *new_flags = 0;                                                                                    \
        khint_t j = 1;                                                                                              \
        {                                                                                                           \
            khint_t t = __ac_HASH_PRIME_SIZE - 1;                                                                   \
            while (__ac_prime_list[t] > new_n_buckets) --t;                                                         \
            new_n_buckets = __ac_prime_list[t+1];                                                                   \
            if (h->size >= (khint_t)(new_n_buckets * __ac_HASH_UPPER + 0.5)) j = 0;                                 \
            else {                                                                                                  \
                new_flags = (uint32_t*)malloc(((new_n_buckets>>4) + 1) * sizeof(uint32_t));                         \
                memset(new_flags, 0xaa, ((new_n_buckets>>4) + 1) * sizeof(uint32_t));                               \
                if (h->n_buckets < new_n_buckets) {                                                                 \
                    h->keys = (khkey_t*)realloc(h->keys, new_n_buckets * sizeof(khkey_t));                          \
                    if (kh_is_map)                                                                                  \
                        h->vals = (khval_t*)realloc(h->vals, new_n_buckets * sizeof(khval_t));                      \
                }                                                                                                   \
            }                                                                                                       \
        }                                                                                                           \
        if (j) {                                                                                                    \
            for (j = 0; j != h->n_buckets; ++j) {                                                                   \
                if (__ac_iseither(h->flags, j) == 0) {                                                              \
                    khkey_t key = h->keys[j];                                                                       \
                    khval_t val;                                                                                    \
                    if (kh_is_map) val = h->vals[j];                                                                \
                    __ac_set_isdel_true(h->flags, j);                                                               \
                    while (1) {                                                                                     \
                        khint_t inc, k, i;                                                                          \
                        k = __hash_func(key);                                                                       \
                        i = k % new_n_buckets;                                                                      \
                        inc = 1 + k % (new_n_buckets - 1);                                                          \
                        while (!__ac_isempty(new_flags, i)) {                                                       \
                            if (i + inc >= new_n_buckets) i = i + inc - new_n_buckets;                              \
                            else i += inc;                                                                          \
                        }                                                                                           \
                        __ac_set_isempty_false(new_flags, i);                                                       \
                        if (i < h->n_buckets && __ac_iseither(h->flags, i) == 0) {                                  \
                            { khkey_t tmp = h->keys[i]; h->keys[i] = key; key = tmp; }                              \
                            if (kh_is_map) { khval_t tmp = h->vals[i]; h->vals[i] = val; val = tmp; }               \
                            __ac_set_isdel_true(h->flags, i);                                                       \
                        } else {                                                                                    \
                            h->keys[i] = key;                                                                       \
                            if (kh_is_map) h->vals[i] = val;                                                        \
                            break;                                                                                  \
                        }                                                                                           \
                    }                                                                                               \
                }                                                                                                   \
            }                                                                                                       \
            if (h->n_buckets > new_n_buckets) {                                                                     \
                h->keys = (khkey_t*)realloc(h->keys, new_n_buckets * sizeof(khkey_t));                              \
                if (kh_is_map)                                                                                      \
                    h->vals = (khval_t*)realloc(h->vals, new_n_buckets * sizeof(khval_t));                          \
            }                                                                                                       \
            free(h->flags);                                                                                         \
            h->flags = new_flags;                                                                                   \
            h->n_buckets = new_n_buckets;                                                                           \
            h->n_occupied = h->size;                                                                                \
            h->upper_bound = (khint_t)(h->n_buckets * __ac_HASH_UPPER + 0.5);                                       \
        }                                                                                                           \
    }                                                                                                               \
    static khint_t kh_put_##name(kh_##name##_t *h, khkey_t key, int *ret)                                    \
    {                                                                                                               \
        khint_t x;                                                                                                  \
        if (h->n_occupied >= h->upper_bound) {                                                                      \
            if (h->n_buckets > (h->size<<1)) kh_resize_##name(h, h->n_buckets - 1);                                 \
            else kh_resize_##name(h, h->n_buckets + 1);                                                             \
        }                                                                                                           \
        {                                                                                                           \
            khint_t inc, k, i, site, last;                                                                          \
            x = site = h->n_buckets; k = __hash_func(key); i = k % h->n_buckets;                                    \
            if (__ac_isempty(h->flags, i)) x = i;                                                                   \
            else {                                                                                                  \
                inc = 1 + k % (h->n_buckets - 1); last = i;                                                         \
                while (!__ac_isempty(h->flags, i) && (__ac_isdel(h->flags, i) || !__hash_equal(h->keys[i], key))) { \
                    if (__ac_isdel(h->flags, i)) site = i;                                                          \
                    if (i + inc >= h->n_buckets) i = i + inc - h->n_buckets;                                        \
                    else i += inc;                                                                                  \
                    if (i == last) { x = site; break; }                                                             \
                }                                                                                                   \
                if (x == h->n_buckets) {                                                                            \
                    if (__ac_isempty(h->flags, i) && site != h->n_buckets) x = site;                                \
                    else x = i;                                                                                     \
                }                                                                                                   \
            }                                                                                                       \
        }                                                                                                           \
        if (__ac_isempty(h->flags, x)) {                                                                            \
            h->keys[x] = key;                                                                                       \
            __ac_set_isboth_false(h->flags, x);                                                                     \
            ++h->size; ++h->n_occupied;                                                                             \
            *ret = 1;                                                                                               \
        } else if (__ac_isdel(h->flags, x)) {                                                                       \
            h->keys[x] = key;                                                                                       \
            __ac_set_isboth_false(h->flags, x);                                                                     \
            ++h->size;                                                                                              \
            *ret = 2;                                                                                               \
        } else *ret = 0;                                                                                            \
        return x;                                                                                                   \
    }                                                                                                               \
    static void kh_del_##name(kh_##name##_t *h, khint_t x)                                                   \
    {                                                                                                               \
        if (x != h->n_buckets && !__ac_iseither(h->flags, x)) {                                                     \
            __ac_set_isdel_true(h->flags, x);                                                                       \
            --h->size;                                                                                              \
        }                                                                                                           \
    }
/* END_MACRO */

/* --- BEGIN OF HASH FUNCTIONS --- */

#define kh_int_hash_func(key) (uint32_t)(key)
#define kh_int_hash_equal(a, b) (a == b)
#define kh_int64_hash_func(key) (uint32_t)((key)>>33^(key)^(key)<<11)
#define kh_int64_hash_equal(a, b) (a == b)

/* chocolateboy 2009-02-13: #ifdef these to avoid warnings about __ac_X31_hash_string being defined but unused */
#ifdef  KH_WANT_STRING_HASH
static khint_t __ac_X31_hash_string(const char *s)
{
    khint_t h = *s;
    if (h) for (++s ; *s; ++s) h = (h << 5) - h + *s;
    return h;
}
#define kh_str_hash_func(key) __ac_X31_hash_string(key)
#define kh_str_hash_equal(a, b) (strcmp(a, b) == 0)
#endif

/* --- END OF HASH FUNCTIONS --- */

/* Other necessary macros... */

#define khash_t(name) kh_##name##_t

/*
 * chocolateboy 2009-02-09
 *
 * make these use the C++ style for templates i.e.
 *
 *     kh_get(type)(table, key)
 *
 * instead of:
 *
 *     kh_get(type, table, key)
 */

/*
#define kh_init(name) kh_init_##name()
#define kh_destroy(name, h) kh_destroy_##name(h)
#define kh_clear(name, h) kh_clear_##name(h)
#define kh_resize(name, h, s) kh_resize_##name(h, s)
#define kh_put(name, h, k, r) kh_put_##name(h, k, r)
#define kh_get(name, h, k) kh_get_##name(h, k)
#define kh_del(name, h, k) kh_del_##name(h, k)
*/

#define kh_init(name) kh_init_##name
#define kh_destroy(name) kh_destroy_##name
#define kh_clear(name) kh_clear_##name
#define kh_resize(name) kh_resize_##name
#define kh_put(name) kh_put_##name
#define kh_get(name) kh_get_##name
#define kh_del(name) kh_del_##name

#define kh_exist(h, x) (!__ac_iseither((h)->flags, (x)))
#define kh_key(h, x) ((h)->keys[x])
#define kh_val(h, x) ((h)->vals[x])
#define kh_value(h, x) ((h)->vals[x])
#define kh_begin(h) (khint_t)(0)
#define kh_end(h) ((h)->n_buckets)
#define kh_size(h) ((h)->size)
#define kh_n_buckets(h) ((h)->n_buckets)

/* More convenient interfaces */

#define KHASH_SET_INIT_INT(name)                                                    \
    KHASH_INIT(name, uint32_t, char, 0, kh_int_hash_func, kh_int_hash_equal)

#define KHASH_MAP_INIT_INT(name, khval_t)                                           \
    KHASH_INIT(name, uint32_t, khval_t, 1, kh_int_hash_func, kh_int_hash_equal)

#define KHASH_SET_INIT_INT64(name)                                                  \
    KHASH_INIT(name, uint64_t, char, 0, kh_int64_hash_func, kh_int64_hash_equal)

#define KHASH_MAP_INIT_INT64(name, khval_t)                                         \
    KHASH_INIT(name, uint64_t, khval_t, 1, kh_int64_hash_func, kh_int64_hash_equal)

typedef const char *kh_cstr_t;
#define KHASH_SET_INIT_STR(name)                                                    \
    KHASH_INIT(name, kh_cstr_t, char, 0, kh_str_hash_func, kh_str_hash_equal)

#define KHASH_MAP_INIT_STR(name, khval_t)                                           \
    KHASH_INIT(name, kh_cstr_t, khval_t, 1, kh_str_hash_func, kh_str_hash_equal)

#endif /* __AC_KHASH_H */


Makefile.PL
use 5.008000;

use strict;
use warnings;

use ExtUtils::MakeMaker;
use ExtUtils::Depends;
use Config;

our $OPTIMIZE;

if ($Config{gccversion}) {
    $OPTIMIZE = '-O3 -Wall -W';
    $OPTIMIZE .= ' -g -Wdeclaration-after-statement' if (-d '.svn');
} elsif ($Config{osname} eq 'MSWin32') {
    $OPTIMIZE = '-O2 -W4';
} else {
    $OPTIMIZE = $Config{optimize};
}

my $package = ExtUtils::Depends->new('B::Hooks::OP::Annotation');

$package->install('hook_op_annotation.h');
$package->add_xs('Annotation.xs');
$package->add_pm('lib/B/Hooks/OP/Annotation.pm' => '$(INST_LIB)/B/Hooks/OP/Annotation.pm');
$package->save_config('Files.pm');

our %MAKEFILE_VARS = $package->get_makefile_vars();

my $clean = $MAKEFILE_VARS{clean};

$clean->{FILES} .= ' Files.pm' if ($clean);

WriteMakefile(
    NAME          => 'B::Hooks::OP::Annotation',
    VERSION_FROM  => 'lib/B/Hooks/OP/Annotation.pm',
    PREREQ_PM     => { 'ExtUtils::Depends' => '0.301' },
    ABSTRACT_FROM => 'lib/B/Hooks/OP/Annotation.pm',
    AUTHOR        => 'chocolateboy <chocolate@cpan.org>',
    LIBS          => [''],
    DEFINE        => '',
    INC           => '-I.',
    FUNCLIST => [
        qw(
          op_annotate
          op_annotation_new
          op_annotation_get
          op_annotation_delete
          op_annotation_group_new
          op_annotation_group_free
        )
    ],
    DL_FUNCS => { 'B::Hooks::OP::Annotation' => [] },
    OPTIMIZE => $OPTIMIZE,
    ($ExtUtils::MakeMaker::VERSION >= 6.31 ? (LICENSE => 'perl') : ()),
    ($ExtUtils::MakeMaker::VERSION >= 6.46 ?
        (META_MERGE => { configure_requires => { 'ExtUtils::Depends' => '0.301' } }) :
        ()
    ),
    %MAKEFILE_VARS
);


hook_op_annotation.h
#ifndef __HOOK_OP_ANNOTATION_H__
#define __HOOK_OP_ANNOTATION_H__

#include "perl.h"

START_EXTERN_C

typedef struct OPAnnotationGroupImpl *OPAnnotationGroup;
typedef  OP *(*OPAnnotationPPAddr)(pTHX);
typedef void (*OPAnnotationDtor)(pTHX_ void *data);

typedef struct {
    OPAnnotationPPAddr op_ppaddr;
    void *data;
    OPAnnotationDtor dtor;
} OPAnnotation;

OPAnnotation * op_annotation_get(OPAnnotationGroup group, OP *op);
OPAnnotation * op_annotation_new(OPAnnotationGroup group, OP *op, void *data, OPAnnotationDtor dtor);
void op_annotatate(OPAnnotationGroup table, OP * op, void *data, OPAnnotationDtor dtor);
void op_annotation_delete(pTHX_ OPAnnotationGroup table, OP * annotation);

OPAnnotationGroup op_annotation_group_new();
void op_annotation_group_free(pTHX_ OPAnnotationGroup group);

END_EXTERN_C

#endif /* __HOOK_OP_ANNOTATION_H__ */


optable.h
/*
 * chocolateboy 2009-02-25
 *
 * This is a customised version of the pointer table implementation in sv.c
 */

#include "ppport.h"

#if PTRSIZE == 8
    /*
     * This is one of Thomas Wang's hash functions for 64-bit integers from:
     * http://www.concentric.net/~Ttwang/tech/inthash.htm
     */
    U32 hash(PTRV u) {
        u = (~u) + (u << 18);
        u = u ^ (u >> 31);
        u = u * 21;
        u = u ^ (u >> 11);
        u = u + (u << 6);
        u = u ^ (u >> 22);
        return (U32)u;
    }
#else
    /*
     * This is one of Bob Jenkins' hash functions for 32-bit integers
     * from: http://burtleburtle.net/bob/hash/integer.html
     */
    U32 hash(PTRV u) {
        u = (u + 0x7ed55d16) + (u << 12);
        u = (u ^ 0xc761c23c) ^ (u >> 19);
        u = (u + 0x165667b1) + (u << 5);
        u = (u + 0xd3a2646c) ^ (u << 9);
        u = (u + 0xfd7046c5) + (u << 3);
        u = (u ^ 0xb55a4f09) ^ (u >> 16);
        return u;
    }
#endif

#define OPTABLE_HASH(op) hash(PTR2nat(op))

typedef struct OPTableEntry {
    struct OPTableEntry *next;
    const OP *key;
    OPAnnotation *value;
} OPTableEntry;

typedef struct OPAnnotationGroupImpl {
    struct OPTableEntry **array;
    UV size;
    UV items;
    NV threshold;
} OPTable;

typedef void (*OPTableEntryValueDtor)(pTHX_ OPAnnotation *);

STATIC OPAnnotation *OPTable_delete(OPTable *table, OP *op);
STATIC OPAnnotation *OPTable_fetch(OPTable *table, const OP *key);
STATIC OPAnnotation * OPTable_store(OPTable *table, const OP *key, OPAnnotation *value);
STATIC OPTableEntry *OPTable_find(OPTable *table, const OP *key);
STATIC OPTable *OPTable_new(UV size, NV threshold);
STATIC void OPTable_clear(pTHX_ OPTable *table, OPTableEntryValueDtor dtor);
STATIC void OPTable_free(OPTable *table);
STATIC void OPTable_grow(OPTable *table);

STATIC OPTable * OPTable_new(UV size, NV threshold) {
    OPTable *table;

    if ((size < 2) || (size & (size - 1))) {
        croak(__PACKAGE__ ": invalid op table size: expected a power of 2 (>= 2), got %u", (unsigned)size);
    }

    if (!((threshold > 0) && (threshold < 1))) {
        croak(__PACKAGE__ ": invalid threshold: expected 0.0 < threshold < 1.0, got %f", threshold);
    }

    Newxz(table, 1, OPTable);

    table->size = size;
    table->threshold = threshold;
    table->items = 0;

    Newxz(table->array, size, OPTableEntry *);

    return table;
}

STATIC OPTableEntry * OPTable_find(OPTable *table, const OP *key) {
    OPTableEntry *entry;
    UV index = OPTABLE_HASH(key) & (table->size - 1);

    for (entry = table->array[index]; entry; entry = entry->next) {
        if (entry->key == key) {
            break;
        }
    }

    return entry;
}

STATIC OPAnnotation *OPTable_delete(OPTable *table, OP *op) {
    OPTableEntry *entry, *prev = NULL;
    OPAnnotation *annotation = NULL;
    UV index = OPTABLE_HASH(op) & (table->size - 1);

    for (entry = table->array[index]; entry; prev = entry, entry = entry->next) {
        if (entry->key == op) {

            if (prev) {
                prev->next = entry->next;
            } else {
                table->array[index] = NULL;
            }

            --(table->items);
            annotation = entry->value;
            Safefree(entry);
            break;
        }
    }

    return annotation;
}

STATIC OPAnnotation * OPTable_fetch(OPTable *table, const OP *key) {
    OPTableEntry const * const entry = OPTable_find(table, key);

    return entry ? entry->value : NULL;
}

STATIC OPAnnotation * OPTable_store(OPTable *table, const OP *key, OPAnnotation *value) {
    OPAnnotation *annotation = NULL;
    OPTableEntry *entry = OPTable_find(table, key);

    if (entry) {
        annotation = entry->value;
        entry->value = value;
    } else {
        const UV index = OPTABLE_HASH(key) & (table->size - 1);
        Newx(entry, 1, OPTableEntry);

        entry->key = key;
        entry->value = value;
        entry->next = table->array[index];

        table->array[index] = entry;
        ++(table->items);

        if (((NV)table->items / (NV)table->size) > table->threshold) {
            OPTable_grow(table);
        }
    }

    return annotation;
}

/* double the size of the array */
STATIC void OPTable_grow(OPTable *table) {
    OPTableEntry **array = table->array;
    const UV oldsize = table->size;
    UV newsize = oldsize * 2;
    UV i;

    Renew(array, newsize, OPTableEntry*);
    Zero(&array[oldsize], newsize - oldsize, OPTableEntry*);
    table->size = newsize;
    table->array = array;

    for (i = 0; i < oldsize; ++i, ++array) {
        OPTableEntry **current_entry_ptr, **entry_ptr, *entry;

        if (!*array) {
            continue;
        }

        current_entry_ptr = array + oldsize;

        for (entry_ptr = array, entry = *array; entry; entry = *entry_ptr) {
            UV index = OPTABLE_HASH(entry->key) & (newsize - 1);

            if (index != i) {
                *entry_ptr = entry->next;
                entry->next = *current_entry_ptr;
                *current_entry_ptr = entry;
                continue;
            } else {
                entry_ptr = &entry->next;
            }
        }
    }
}

STATIC void OPTable_clear(pTHX_ OPTable *table, OPTableEntryValueDtor dtor) {
    if (table && table->items) {
        OPTableEntry ** const array = table->array;
        UV riter = table->size - 1;

        do {
            OPTableEntry *entry = array[riter];

            while (entry) {
                OPTableEntry * const temp = entry;
                entry = entry->next;
                dtor(aTHX_ temp->value);
                Safefree(temp);
            }

            /* chocolateboy 2008-01-08
             *
             * make sure we clear the array entry, so that subsequent probes fail
             */

            array[riter] = NULL;
        } while (riter--);

        table->items = 0;
    }
}

STATIC void OPTable_free(OPTable *table) {
    if (table) {
        Safefree(table);
    }
}
